src/ed25519.rs

```diff
@@ -727,36 +719,27 @@ impl PublicKey {
     ///
     /// # Return
     ///
-    /// Returns true if the signature was successfully verified, and
-    /// false otherwise.
-    pub fn verify<D>(&self, message: &[u8], signature: &Signature) -> bool
+    /// Returns `Ok(())` if the signature is valid, and `Err` otherwise.
+    #[allow(non_snake_case)]
+    pub fn verify<D>(&self, message: &[u8], signature: &Signature) -> Result<(), SignatureError>
             where D: Digest<OutputSize = U64> + Default
     {
-        let mut h: D = D::default();
-        let mut a: EdwardsPoint;
-        let ao:  Option<EdwardsPoint>;
-        let mut digest: [u8; 64] = [0u8; 64];
-
-        ao = self.decompress();
-
-        if ao.is_some() {
-            a = ao.unwrap();
-        } else {
-            return false;
-        }
-        a = -(&a);
```

Comment:
Why did you change to negating `k` instead?

Reply:
No real reason, I wanted to avoid a point operation, but negating the Edwards point is probably cheaper (since it doesn't require a reduction mod L)

Reply:
I benched scalar and Edwards point negation and point negation is like twice as fast (they're both in the tens of nanoseconds on my machine so not that it really matters but).